azure/src/main/java/org/apache/iceberg/azure/adlsv2/VendedAdlsCredentialProvider.java

```diff
@@ -0,0 +1,197 @@
+    if (refresherForAccountMap.containsKey(storageAccount)) {
+      return refresherForAccountMap.get(storageAccount).azureSasCredential();
+    } else {
+      AzureSasCredentialRefresher azureSasCredentialRefresher =
+          new AzureSasCredentialRefresher(
+              () -> this.sasTokenWithExpiration(storageAccount), credentialRefreshExecutor());
+      refresherForAccountMap.put(storageAccount, azureSasCredentialRefresher);
+      return azureSasCredentialRefresher.azureSasCredential();
+    }
+  }
+
+  private Pair<String, Long> sasTokenWithExpiration(String storageAccount) {
+    LoadCredentialsResponse response = fetchCredentials();
+    List<Credential> adlsCredentials =
+        response.credentials().stream()
+            .filter(c -> c.prefix().contains(storageAccount))
+            .collect(Collectors.toList());
+    Preconditions.checkState(
+        !adlsCredentials.isEmpty(),
+        String.format("Invalid ADLS Credentials for storage-account %s: empty", storageAccount));
+    Preconditions.checkState(
+        adlsCredentials.size() == 1,
+        "Invalid ADLS Credentials: only one ADLS credential should exist per storage-account");
+
+    Credential adlsCredential = adlsCredentials.get(0);
+    checkCredential(adlsCredential, AzureProperties.ADLS_SAS_TOKEN_PREFIX + storageAccount);
+    checkCredential(
+        adlsCredential, AzureProperties.ADLS_SAS_TOKEN_EXPIRES_AT_MS_PREFIX + storageAccount);
+
+    String updatedSasToken =
+        adlsCredential.config().get(AzureProperties.ADLS_SAS_TOKEN_PREFIX + storageAccount);
+    Long tokenExpiresAtMillis =
+        Long.parseLong(
+            adlsCredential
+                .config()
+                .get(AzureProperties.ADLS_SAS_TOKEN_EXPIRES_AT_MS_PREFIX + storageAccount));
+
+    return Pair.of(updatedSasToken, tokenExpiresAtMillis);
+  }
+
+  private Map<String, AzureSasCredentialRefresher> azureSasCredentialRefresherMap() {
+    if (this.azureSasCredentialRefresherMap == null) {
+      synchronized (this) {
+        if (this.azureSasCredentialRefresherMap == null) {
+          this.azureSasCredentialRefresherMap = Maps.newHashMap();
+        }
+      }
+    }
+    return this.azureSasCredentialRefresherMap;
+  }
+
+  private ScheduledExecutorService credentialRefreshExecutor() {
+    if (this.refreshExecutor == null) {
+      synchronized (this) {
+        if (this.refreshExecutor == null) {
+          this.refreshExecutor = ThreadPools.newScheduledPool(THREAD_PREFIX, 1);
+        }
+      }
+    }
+    return this.refreshExecutor;
+  }
+
+  private RESTClient httpClient() {
+    if (null == client) {
+      synchronized (this) {
+        if (null == client) {
+          client = HTTPClient.builder(properties).uri(properties.get(URI)).build();
+        }
+      }
+    }
+
+    return client;
+  }
+
+  private LoadCredentialsResponse fetchCredentials() {
+    Map<String, String> headers =
+        RESTUtil.merge(
+            configHeaders(properties),
```

Comment:
can you elaborate why we need the config headers here? I believe the auth headers should be enough

Reply:
Tried to implement the similar behaviour present in [RESTSessionCatalog](https://github.com/apache/iceberg/blob/38c8daa4eae8a75ab46571f1efce1609100f53dd/core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java#L262) for loadTable(), where catalog can be configured to pass explicit headers to server by setting the configuration with header. prefix.

While I was testing these changes with vended credentials from loadTable response (workaround to mimic new credential endpoint), I saw the possibility that REST server implementations may use headers as feature rollout gates/ additional metadata to decide credential expiration time. Similarly new credential endpoint may also support accepting headers from client as additional metadata.

Reply:
I think for now let's leave the config headers out, because we also don't have them for vended S3/GCS credential providers

Reply:
Sure, will remove them.

Reply:
Removed the config headers.